File: src/com/uipkg/simple/UIFactory.java
package com.uipkg.simple;

import java.awt.event.AWTEventListener;
import java.awt.event.WindowListener;
import java.awt.event.KeyListener;
import java.awt.event.FocusListener;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.awt.event.MouseWheelListener;
import java.awt.event.ComponentListener;

import java.util.ArrayList;

import javax.swing.JComponent;
import javax.swing.JFrame;
import javax.swing.WindowConstants;

import com.uipkg.components.Configurator;
import com.uipkg.components.defined.DefaultCFG;
import com.uipkg.exceptions.NonApplicableException;

public final class UIFactory extends JFrame implements SimpleFrame {
  private static final long serialVersionUID = 1L;

  public UIFactory(Configurator cfg, JComponent... components) {
    setPreferredSize(cfg.getDimension());
    setBackground(cfg.getBackgroundColor());
    setForeground(cfg.getForegroundColor());
    setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
    setTitle(cfg.getTitleStr());
    if (cfg.getIcon() != null) {
      setIconImage(cfg.getIcon().getImage());
    }
    for (JComponent component : components) {
      add(component);
    }
    ArrayList<AWTEventListener> listeners = cfg.getListenerConfigurator().getListeners();
    for (AWTEventListener listener : listeners) {
      if (listener instanceof WindowListener) {
        addWindowListener((WindowListener) listener);
      } else if (listener instanceof KeyListener) {
        addKeyListener((KeyListener) listener);
      } else if (listener instanceof FocusListener) {
        addFocusListener((FocusListener) listener);
      } else if (listener instanceof MouseListener) {
        addMouseListener((MouseListener) listener);
      } else if (listener instanceof MouseMotionListener) {
        addMouseMotionListener((MouseMotionListener) listener);
      } else if (listener instanceof MouseWheelListener) {
        addMouseWheelListener((MouseWheelListener) listener);
      } else if (listener instanceof ComponentListener) {
        addComponentListener((ComponentListener) listener);
      } else {
        try {
          throw new NonApplicableException("Listener type not supported");
        } catch (NonApplicableException e) {
          e.printStackTrace();
        }
      }
    }
  }

  public UIFactory() {
    this(DefaultCFG.CFG_480P);
  }

  public UIFactory(JComponent... components) {
    this(DefaultCFG.CFG_480P, components);
  }

  @Override
  public void run() {
    pack();
    setVisible(true);
  }

  @Override
  public void runWithTask(Runnable s) throws NonApplicableException {
    throw new NonApplicableException("UIFactory frames cannot be attached with a synchronous task.");
  }

  @Override
  public void killTask() throws NonApplicableException {
    throw new NonApplicableException("UIFactory frames cannot be attached with a synchronous task.");
  }

  @Override
  public JFrame getInstanceFrame() {
    return this;
  }
}

File: src/com/uipkg/simple/UIFactory.kt
package com.uipkg.simple

import com.uipkg.components.Configurator
import com.uipkg.components.defined.DefaultCFG
import com.uipkg.exceptions.NonApplicableException
import java.awt.event.*
import javax.swing.JComponent
import javax.swing.JFrame
import kotlin.Throws
import java.lang.Runnable
import javax.swing.WindowConstants

class UIFactory(cfg: Configurator, vararg components: JComponent?) : JFrame(), SimpleFrame {
    constructor() : this(DefaultCFG.CFG_480P) {}
    constructor(vararg components: JComponent?) : this(DefaultCFG.CFG_480P, *components) {}

    override fun run() {
        pack()
        isVisible = true
    }

    @Throws(NonApplicableException::class)
    override fun runWithTask(s: Runnable) {
        throw NonApplicableException("UIFactory frames cannot be attached with a synchronous task.")
    }

    @Throws(NonApplicableException::class)
    override fun killTask() {
        throw NonApplicableException("UIFactory frames cannot be attached with a synchronous task.")
    }

    override fun getInstanceFrame(): JFrame {
        return this
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        preferredSize = cfg.dimension
        background = cfg.backgroundColor
        foreground = cfg.foregroundColor
        defaultCloseOperation = DISPOSE_ON_CLOSE
        title = cfg.titleStr
        if (cfg.icon != null) {
            iconImage = cfg.icon.image
        }
        for (component in components) {
            add(component)
        }
        val listeners = cfg.listenerConfigurator.listeners
        for (listener in listeners) {
            if (listener is WindowListener) {
                addWindowListener(listener as WindowListener)
            } else if (listener is KeyListener) {
                addKeyListener(listener as KeyListener)
            } else if (listener is FocusListener) {
                addFocusListener(listener as FocusListener)
            } else if (listener is MouseListener) {
                addMouseListener(listener as MouseListener)
            } else if (listener is MouseMotionListener) {
                addMouseMotionListener(listener as MouseMotionListener)
            } else if (listener is MouseWheelListener) {
                addMouseWheelListener(listener as MouseWheelListener)
            } else if (listener is ComponentListener) {
                addComponentListener(listener as ComponentListener)
            } else {
                try {
                    throw NonApplicableException("Listener type not supported")
                } catch (e: NonApplicableException) {
                    e.printStackTrace()
                }
            }
        }
    }
}